### DISCUSSION TITLE
Questions about files, code, and node management

### DISCUSSION THREAD

**Discussion by aVersionOfReality** [bnpr/Malt#129] [Q&A]
Some questions related to setting things up. Trying to figure out exactly what my options are and how I'm supposed to be doing things.

1) I've made a shader setup using nodes, and I see that I can take the .mesh.glsl file outputted from that and load it back in as the shader source instead of using the node tree. However, when I do this, it seems all the values are set to defaults of either 0 or 1. Is there a way to define what they should default to in the .mesh.glsl file? Or better yet, in the nodes that its generated from? (I have noticed that naming the nodes does translate to the property names, so that works.)

2) Is there a way to control which node inputs show up in the materials panel and which don't? How about if bringing the file in as a shader source? I'm trying to make essentially a big uber shader. It'll be unusable if every single input on any node is going to show up.

3) I have figured out how to add my own custom node_utils folder and custom utils. That's all working fine. Can I do something similar with a .mesh.glsl shader generated from nodes? Can that be brought back in as its own node somehow? (This would effectively be a node group.) I have tried including the file the same way you include node_utils files and it causes blender to freeze. But there's a lot going on in the .mesh.glsl files created from nodes.

4) If 3 can't be done, what is the proper approach to adding a new shader node? My shader is essentially a scene_diffuse * scene_toon * AO with some math inserted to adjust the strength of some of these. It is all existing things, just combined together with math. I want to get this as a single node with only the inputs I need. I assume if I look through the existing sample shader files, I can copy and paste bits together to build this up. But its already getting put together in the .mesh.glsl file being generated from these nodes, so it seems like it should be easier if I just knew what part of that is needed, what parts of it I can remove, and what else needs to be included or not?
![image](https://user-images.githubusercontent.com/36803956/133691399-63e20dc1-ed57-42bc-927c-7176e259163c.png)

5) Can I add a util node that is just a list of vec3's with no input, and each vec3 being its own output? Or return an entry in the list based on an index? The values of the vec3s would be defined in the file. The current schema I can see for existing utils always is multiple inputs, but only a single output. (This would be used for bringing in a color palette that can be used in multiple different materials in the project.)

6) There are a large amount of nodes that don't seem to work, or I can't figure out how to get the input they seem to need. For example:
![image](https://user-images.githubusercontent.com/36803956/133693525-b55e6870-09cc-48cc-8e62-1515a351dfc9.png)
These nodes seem to want image textures as inputs? My understanding is that these are components used elsewhere, and we're not actually intended to be using them in the node editor, and instead access these functions through different nodes. Is that correct?

7) Is every material's shader compiled separately, even if they are using the same shader source file or node tree?

Thank you!

**Comment by pragma37** [comment]
1. Nope, because default values are stored in the nodes themselves. Some properties (like textures and gradients) can't be stored in GLSL files anyway. Why do you want to use the generated code directly instead of the node tree?

2. Not yet, but it's on my to-do list, it's a common request. That said, if you have an uber-shader that you'll use very often, writing it as GLSL is probably not a bad idea.

3. Yes, but it requires some manual work.
Basically, you have to take the NODES_COMMON_PIXEL_SHADER function, rename it to your new node name, and put the uniform parameters as function parameters.

So if you have, for example:

```glsl
#include "Pipelines/NPR_Pipeline.glsl"
#include "Node Utils/node_utils.glsl"

#ifdef PIXEL_SHADER
#ifdef MAIN_PASS
layout (location = 3) out vec4 OUT_CustomA;
layout (location = 4) out vec4 OUT_CustomB;
layout (location = 5) out vec4 OUT_CustomC;
layout (location = 6) out vec4 OUT_CustomD;

#endif
#endif

void NODES_COMMON_PIXEL_SHADER(Surface S, inout PixelOutput PO , out vec4 CustomA, out vec4 CustomB, out vec4 CustomC, out vec4 CustomD);

void COMMON_PIXEL_SHADER(Surface S, inout PixelOutput PO)
{
    vec4 CustomA = vec4(0);
vec4 CustomB = vec4(0);
vec4 CustomC = vec4(0);
vec4 CustomD = vec4(0);


    NODES_COMMON_PIXEL_SHADER(S, PO , CustomA, CustomB, CustomC, CustomD);

    #ifdef PIXEL_SHADER
    #ifdef MAIN_PASS
    {
        OUT_CustomA = CustomA;
OUT_CustomB = CustomB;
OUT_CustomC = CustomC;
OUT_CustomD = CustomD;

    }
    #endif
    #endif
}



uniform vec4 U_0_COMMON_PIXEL_SHADEROutput_0_CustomA;
uniform vec4 U_0_COMMON_PIXEL_SHADEROutput_0_CustomB;
uniform vec4 U_0_COMMON_PIXEL_SHADEROutput_0_CustomC;
uniform vec4 U_0_COMMON_PIXEL_SHADEROutput_0_CustomD;
uniform float U_0_get_toon_0_size;
uniform float U_0_get_toon_0_gradient_size;
uniform float U_0_get_toon_0_specularity;
uniform float U_0_get_toon_0_offset;
uniform float U_0_vec4_from_vec3_0_a;


void NODES_COMMON_PIXEL_SHADER(Surface S, inout PixelOutput PO , out vec4 CustomA, out vec4 CustomB, out vec4 CustomC, out vec4 CustomD)
{

	vec3 _get_toon_0_result = get_toon(U_0_get_toon_0_size,U_0_get_toon_0_gradient_size,U_0_get_toon_0_specularity,U_0_get_toon_0_offset);

	vec4 _vec4_from_vec3_0_result = vec4_from_vec3(_get_toon_0_result,U_0_vec4_from_vec3_0_a);

	PixelOutput _PixelOutput = PO;
	_PixelOutput.color = _vec4_from_vec3_0_result;

	PO = _PixelOutput;
	CustomA = U_0_COMMON_PIXEL_SHADEROutput_0_CustomA;
	CustomB = U_0_COMMON_PIXEL_SHADEROutput_0_CustomB;
	CustomC = U_0_COMMON_PIXEL_SHADEROutput_0_CustomC;
	CustomD = U_0_COMMON_PIXEL_SHADEROutput_0_CustomD;

}
```

You remove the mesh shader setup code:

```glsl
uniform vec4 U_0_COMMON_PIXEL_SHADEROutput_0_CustomA;
uniform vec4 U_0_COMMON_PIXEL_SHADEROutput_0_CustomB;
uniform vec4 U_0_COMMON_PIXEL_SHADEROutput_0_CustomC;
uniform vec4 U_0_COMMON_PIXEL_SHADEROutput_0_CustomD;
uniform float U_0_get_toon_0_size;
uniform float U_0_get_toon_0_gradient_size;
uniform float U_0_get_toon_0_specularity;
uniform float U_0_get_toon_0_offset;
uniform float U_0_vec4_from_vec3_0_a;


void NODES_COMMON_PIXEL_SHADER(Surface S, inout PixelOutput PO , out vec4 CustomA, out vec4 CustomB, out vec4 CustomC, out vec4 CustomD)
{

	vec3 _get_toon_0_result = get_toon(U_0_get_toon_0_size,U_0_get_toon_0_gradient_size,U_0_get_toon_0_specularity,U_0_get_toon_0_offset);

	vec4 _vec4_from_vec3_0_result = vec4_from_vec3(_get_toon_0_result,U_0_vec4_from_vec3_0_a);

	PixelOutput _PixelOutput = PO;
	_PixelOutput.color = _vec4_from_vec3_0_result;

	PO = _PixelOutput;
	CustomA = U_0_COMMON_PIXEL_SHADEROutput_0_CustomA;
	CustomB = U_0_COMMON_PIXEL_SHADEROutput_0_CustomB;
	CustomC = U_0_COMMON_PIXEL_SHADEROutput_0_CustomC;
	CustomD = U_0_COMMON_PIXEL_SHADEROutput_0_CustomD;

}
```

Rename the function and move the uniforms to the parameters list:


```glsl
void My_Custom_Node(Surface S, inout PixelOutput PO , out vec4 CustomA, out vec4 CustomB, out vec4 CustomC, out vec4 CustomD,
vec4 U_0_COMMON_PIXEL_SHADEROutput_0_CustomA,
vec4 U_0_COMMON_PIXEL_SHADEROutput_0_CustomB,
vec4 U_0_COMMON_PIXEL_SHADEROutput_0_CustomC,
vec4 U_0_COMMON_PIXEL_SHADEROutput_0_CustomD,
float U_0_get_toon_0_size,
float U_0_get_toon_0_gradient_size,
float U_0_get_toon_0_specularity,
float U_0_get_toon_0_offset,
float U_0_vec4_from_vec3_0_a)
{

	vec3 _get_toon_0_result = get_toon(U_0_get_toon_0_size,U_0_get_toon_0_gradient_size,U_0_get_toon_0_specularity,U_0_get_toon_0_offset);

	vec4 _vec4_from_vec3_0_result = vec4_from_vec3(_get_toon_0_result,U_0_vec4_from_vec3_0_a);

	PixelOutput _PixelOutput = PO;
	_PixelOutput.color = _vec4_from_vec3_0_result;

	PO = _PixelOutput;
	CustomA = U_0_COMMON_PIXEL_SHADEROutput_0_CustomA;
	CustomB = U_0_COMMON_PIXEL_SHADEROutput_0_CustomB;
	CustomC = U_0_COMMON_PIXEL_SHADEROutput_0_CustomC;
	CustomD = U_0_COMMON_PIXEL_SHADEROutput_0_CustomD;

}
```

And probably return just the _PixelOutput.color or whatever you are calculating in your nodes and remove unused stuff:

```glsl
vec4 My_Custom_Node(Surface S, in PixelOutput PO,
float U_0_get_toon_0_size,
float U_0_get_toon_0_gradient_size,
float U_0_get_toon_0_specularity,
float U_0_get_toon_0_offset,
float U_0_vec4_from_vec3_0_a)
{

	vec3 _get_toon_0_result = get_toon(U_0_get_toon_0_size,U_0_get_toon_0_gradient_size,U_0_get_toon_0_specularity,U_0_get_toon_0_offset);

	vec4 _vec4_from_vec3_0_result = vec4_from_vec3(_get_toon_0_result,U_0_vec4_from_vec3_0_a);

	PixelOutput _PixelOutput = PO;
	_PixelOutput.color = _vec4_from_vec3_0_result;
	
	return _PixelOutput.color;
}
```
4. Just copy-paste the function into your node_utils file and that's it.

5. Yes, in GLSL, in addition to the parameter result type. For example:
```glsl
vec4 red() { return vec4(1,0,0,1); }
```
You can also do this:
```glsl
void color(out vec4 red) { red = vec4(1,0,0,1); }
```
(Notice the *out* keyword)
And this:
```glsl
void colors(out vec4 red, out vec4 green, out vec4 blue)
{
    red = vec4(1,0,0,1);
    green = vec4(0,1,0,1);
    blue = vec4(0,0,1,1);
}
```
You can also use *inout* if you want to get a variable both as input and output:
```glsl
void brighter(inout vec4 color) { color.rgb += vec3(0.1, 0.1, 0.1); }
```

6. Explained here:
#127
These can actually be useful in the node editor for advanced use-cases.

7. Nope, each shader source (and, by extension, node tree) is only compiled once and shared by all the materials that use it.

**Reply by aVersionOfReality** [reply]
1&2: I'm not sure that I do want to use it directly, I'm just curious as to how it all works. And to clarify, when you say nope, you mean the defaults cannot be saved from the nodes? Defaults can be setup in the GLSL file itself, right? (If not, where are the various sample shader defaults coming from?

3: Great, got it working! Thanks for putting that much detail into this one, or I wouldn't have gotten it.

4: No problems with that either.

5: Also got it working! On your _inout_ example, it looks like when added as a node, it just defaults it to White though? Is it supposed to have the inputted color be set to the color defined in the code? Or am I misunderstanding?

7: So if I make myself an uber shader, bring it in as its own node, and then have every material be that node but with various different inputs in their own node tree, its still only one compile for that uber shader node? If so, that is a huge advantage over Eevee!

Thanks for all the answers!

**Reply by pragma37** [reply]
1&2: As I said, not every default can be stored directly in GLSL (like textures). But even for the ones where it's possible, the current shader generation code doesn't support it.

5. It returns a slightly brighter (+ 0.1) color than the one you input. So if you input (0.5,0.5,0.5), it returns (0.6,0.6,0.6).

7. Nope, only materials that share the exact same NodeTree/GLSL File share the same compiled shader.
